### PR TITLE
MatrixGate names don't survive serialization

### LINKED ISSUE #5999: `MatrixGate` names don't survive serialization
**Description of the issue**

`_name` is excluded from `MatrixGate._json_dict_()` and so is reset to `None` by serialization+deserialization

**How to reproduce the issue**

```python
gate = cirq.MatrixGate(np.eye(2), name="name")
gate_after_serialization = cirq.read_json(json_text=cirq.to_json(gate))

print(gate._name)  # "name"
print(gate_after_serialization._name)  # None
```

**Cirq version**

```
1.2.0.dev20230105212249
```

### PR DESCRIPTION
**Pull request by markedmiston**
Closes #5999

### CODE REVIEW

File: cirq-core/cirq/ops/matrix_gates.py [hunk 1]
```diff
@@ -114,11 +114,11 @@ def with_name(self, name: str) -> 'MatrixGate':
         return MatrixGate(self._matrix, name=name, qid_shape=self._qid_shape, unitary_check=False)
 
     def _json_dict_(self) -> Dict[str, Any]:
-        return {'matrix': self._matrix.tolist(), 'qid_shape': self._qid_shape}
+        return {'matrix': self._matrix.tolist(), 'qid_shape': self._qid_shape, **({'name': self._name} if self._name else {})}
```
Comment (viathor) [inline cirq-core/cirq/ops/matrix_gates.py:117]: Strictly speaking this sends empty string to `None` after serialization+deserialization.

**Review by viathor — APPROVED** [review]
LGTM with a minor comment.

File: cirq-core/cirq/ops/matrix_gates.py (hunk 1 above)
Comment (markedmiston) [reply cirq-core/cirq/ops/matrix_gates.py:117]: Do we consider an empty string as a valid name? If so, I’ll update the check to more explicitly check None.

Comment (viathor) [reply cirq-core/cirq/ops/matrix_gates.py]: Well, at present the empty string is accepted by the ctor, so let's make sure it works in serde, too.

Comment (markedmiston) [reply cirq-core/cirq/ops/matrix_gates.py]: Good point. I've updated the check logic and added an empty name test.

**Pull request merged**